Update send_produce_error to decode failed_messages with built-in decoder.

DeeChau opened the pull request:
## Description

Fixes #19 - Update `send_produce_error` to decode `failed_messages` with built-in decoder. 

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?

Tested on a staging environment of fadmin. Metrics show up in datadog tracking when errors happen during the delivery of messages to kafka. i.e 
```
Failed to send all messages to multi.Flyers.FlyerDataResponse/2; attempting retry 1 of 10 after 5s
Failed to fetch metadata from kafka://kafka-stg.bootstrap.wishabi.net:9093: Connection error Errno::ETIMEDOUT: Connection timed out
```

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

dorner commented on lib/deimos/instrumentation.rb:
```diff
@@ -48,11 +48,17 @@ def self.send_produce_error(event)
+        decoder = Deimos::KafkaMessage.decoder(topic)
+        payloads = batch.map do |m|
+          decoder&.decoder&.decode(m.value)
+        end
```
I think this is a bit convoluted. There's an easier way to do this now:
```ruby
decoder = Deimos.schema_backend(schema: producer.config[:schema],
                                           namespace: producer.config[:namespace])
payloads = batch.map { |m| decoder.decode(m.value) }
```
Also don't forget to handle the case where the producer wasn't found.

DeeChau commented:
Addressed comments and tested on a staging environment. Metrics show up in datadog tracking when errors happen during the delivery of messages to kafka. e.g.
```
Failed to send all messages to multi.Flyers.FlyerDataResponse/2; attempting retry 1 of 10 after 5s
Failed to fetch metadata from kafka://kafka-stg.bootstrap.wishabi.net:9093: Connection error Errno::ETIMEDOUT: Connection timed out
```
have corresponding metrics in received.
![image](https://user-images.githubusercontent.com/11283382/89421812-833de700-d702-11ea-9514-359d1aa8bf5c.png)

**CloudWatch Logs Insights**  
region: us-east-1  
log-group-names: fadmin-multi  
start-time: 2020-08-04T18:31:31.000Z  
end-time: 2020-08-04T20:50:23.000Z  
query-string:
```
filter @message LIKE /Failed to send all messages to/ 
| filter @message LIKE /keeping remaining messages in buffer/
| sort @timestamp desc
| parse '[7150]: Failed to send all messages to */' as topic
| stats count() by topic
```
--------------------------------------------
|             topic              | count() |
|--------------------------------|---------|
| multi.Flyers.FlyerFsas         | 12      |
| multi.Flyers.FlyerDataResponse | 10      |
| multi.fadmin_flyers            | 8       |
--------------------------------------------

DeeChau commented on lib/deimos/instrumentation.rb:
```diff
@@ -46,13 +46,18 @@ def self.send_produce_error(event)
       messages.group_by(&:topic).each do |topic, batch|
-        next if batch.empty?
+        producer = Deimos::Producer.descendants.find { |c| c.topic == topic }
+        next unless batch.any? && producer
```
sure thing

DeeChau commented:
@dorner Made the changes!

dorner commented:
Nice, thanks!
